### PR TITLE
@bentley/imodeljs-native 3.5.8

### PR DESCRIPTION
**Pull request by nick4598**

### CODE REVIEW

File: full-stack-tests/backend/package.json [hunk 1]
```diff
@@ -40,7 +40,7 @@
     "@bentley/cif-subsurface-schema": "^1.0.3",
     "@bentley/cif-subsurface-conflict-analysis-schema": "^1.0.3",
     "@bentley/cif-units-schema": "^1.0.1",
-    "@bentley/imodeljs-native": "3.5.6",
+    "@bentley/imodeljs-native": "3.5.8",
```
Comment (aruniverse) [inline full-stack-tests/backend/package.json:43]: never thought about it til now, but why does this full-stack-test have a direct-dep on imjs-native?
shoudlnt this be a transitive dep via core-backend?

Comment (nick4598) [reply full-stack-tests/backend/package.json:43]: My initial guess is that we used to be importing something from imodeljs-native directly, something that wasn't re-exported by the core-backend.ts file. Maybe it was removed in the last few months and this is no longer necessary.

Comment (aruniverse) [reply full-stack-tests/backend/package.json:43]: could you remove this dep, and lets see if anything breaks? i doubt it will but would be goof to know

**Review by aruniverse — APPROVED** [review]

File: full-stack-tests/backend/package.json (hunk 1 above)
Comment (nick4598) [reply full-stack-tests/backend/package.json]: I updated the wiki to no longer call out full-stack-tests and removed the line from the .sh script which automatically updates the package versions

**Review by calebmshafer — APPROVED** [review]

**Review by tcobbs-bentley — APPROVED** [review]

**Comment by nick4598** [comment]
Lots of test failures with "BE_SQLITE_ERROR: error importing schema". @ColinKerr could this be related to the other schema issues we were seeing? 

I won't be able to look into this until monday, because I'm out of office tomorrow.

**Comment by nick4598** [comment]
> Lots of test failures with "BE_SQLITE_ERROR: error importing schema". @ColinKerr could this be related to the other schema issues we were seeing?
> 
> I won't be able to look into this until monday, because I'm out of office tomorrow.

I believe I see two issues here.. The first is that the BisCore schema cannot be located. 
>Error | ECObjectsNative | Unable to locate referenced schema BisCore.01.00.15 while deserializing PresentationRules.01.00.01

The BisCore schema is located in Assets/ECSchemas/Dgn.

At the time of searching for the referenced BisCore schema in the code I see 4 schema locaters in the ECSchemaReadContext class... one for the cache of already seen schemas, one which checks the iModel itself for the presence of the schema and then two (one of them has two search paths) which search Assets/*, Assets/ECSchemas/Standard/*, and Assets/ECSchemas/Domain/*. 

Originally I thought that the answer was to add another search path Assets/ECSchemas/Dgn... but I'm not sure its this simple. I do see the biscore schema get loaded from that folder at some point in the _OnDbOpened I believe, but its just not available at the time of attempting to locate the referenced schema. It seems that the ECSchemaReadContext constructor is called many times(expected behavior I assume), which is why the referenced schema is unable to be located. This part I'm still trying to fully understand.

The second issue is one I discovered when I put a copy of BisCore.ecschema in one of those above directories, 
>Error | ECDb | Failed to import schema 'BisCore.01.00.15'. Current ECDb profile version (4.0.0.1) only support schemas with EC version < 3.2. ECDb profile version upgrade is required to import schemas with EC Version >= 3.2.

I'm guessing this boils down to having to open the seed files with the appropriate open options to get the profile version to upgrade. Is that a manual process or is there a script located somewhere for this purpose?

**Comment by nick4598** [comment]
After talking with Colin...

The first issue, looks like we'll have to add the Dgn directory to the list of searchPaths for the schema locaters. That will require a new PR into the imodel-native code. Instead of abandoning this PR, we could temporarily get around this and still merge 3.5.8 if we open the seedfiles OR the testfiles created off of the seedfiles, with DomainOptions.Upgrade.

The above will store the BisCore 01.00.15 in the iModel that the PresentationRules.ecschema .xml is looking for. That way the schema doesn't have to be found in one of the searchPaths. I'm not sure if this is a good option or not... but regardless we will need to update the Profile on either the seed or the test db created off of the seed for grigas's tests to pass ( this fixes the second issue I mentioned in my longer post) .

@grigasp Any preference on whether we update the seed or the test file created off of the seed?

**Comment by grigasp** [comment]
> After talking with Colin...
> 
> The first issue, looks like we'll have to add the Dgn directory to the list of searchPaths for the schema locaters. That will require a new PR into the imodel-native code. Instead of abandoning this PR, we could temporarily get around this and still merge 3.5.8 if we open the seedfiles OR the testfiles created off of the seedfiles, with DomainOptions.Upgrade.
> 
> The above will store the BisCore 01.00.15 in the iModel that the PresentationRules.ecschema .xml is looking for. That way the schema doesn't have to be found in one of the searchPaths. I'm not sure if this is a good option or not... but regardless we will need to update the Profile on either the seed or the test db created off of the seed for grigas's tests to pass ( this fixes the second issue I mentioned in my longer post) .
> 
> @grigasp Any preference on whether we update the seed or the test file created off of the seed?

The third option was to just use an empty iModel for these tests - I pushed the change.

**Review by saskliutas — APPROVED** [review]

**Comment by pmconne** [comment]
@ColinKerr please respond here regarding questions about sudden proliferation of schema version issues.
For a new example, see [here](https://github.com/iTwin/itwinjs-core/pull/4637/commits/e24d41d43167d349db8803f4643551a565a8c0f8).
Will consumers of our packages be impacted? Is this desirable/expected?

**Review by MichaelBelousov — APPROVED** [review]

**Comment by ColinKerr** [comment]
> @ColinKerr please respond here regarding questions about sudden proliferation of schema version issues.
> For a new example, see [here](https://github.com/iTwin/itwinjs-core/pull/4637/commits/e24d41d43167d349db8803f4643551a565a8c0f8).
> Will consumers of our packages be impacted? Is this desirable/expected?

The latest schema changes included the following things
1. Updating core schemas to EC 3.2
2. Adding the new BisCustomAttributes schema as a reference to many schemas
3. Updating schema reference versions

If anyone thinks there is another type of issue happening please let me know so we can get an handle on it.

Updating core schemas to EC 3.2 means that the iModel must be off profile 4.0.0.2 or greater.  Connectors have been set to upgrade the profile for 2+ years but many of our test files were still on the older profile.  So it is unlikely that a user has any actual files that are still on the older profile.  I do not think this will cause major user production issues though there may be some problems if code does not know how to manage profile upgrades.  If they do not handle profile upgrades correctly other already released schemas could also cause the change, this just increases the chance they will hit the issue.  The linked commit appears to be handling an old seed file.

Adding the new BisCustomAttributes schema as a reference.  This exposed a bug in the C++ logic that handles schema upgrades for domains which have C++ handlers.  We held back BisCore, Generic, Functional and PhysicalMaterial schemas.  It can cause failure to open an iModel in readonly mode.  We are working on a fix but we should hold back any other schemas if they have handlers registered in iTwin.js until then.  If we miss a schema I could see this causing user issues.  Connectors are going out with everything but BisCore and Generic as they will not face the issue (they upgrade schemas on open) and it's safe once the schemas are in the iModels.

Updating schema reference versions caused the failure in presentation tests and exposed how picky and inconsistent our schema import logic is.  There should be an in progress fix to improve the schema import logic by registering all standard schema paths when importing a schema from disk or xml string.  It will still require schemas to be imported in reference order in some cases.  We can and should add the ability to supply custom directories to search for reference schemas, but this is an independent change.  I believe this may cause some user issues with code that does schema imports until we have the basic fix in place on the C++ side.


We have hit several issues independent of this that also indicate that we need to have better testing for old profiles and iModels that have been updated over time.  This issue rash of issues has shown that we need to do better on tests that happen before we push new schemas.  Several improvements are in progress to improve the existing iModel Evolution tests and extend testing to include testing iTwin.js with new schemas before they are pushed.  In the longer term we will have to do more but the exact steps are still under discussion.

**Comment by pmconne** [comment]
Thanks @ColinKerr.
> I do not think this will cause major user production issues though there may be some problems.

Any potential problems need to be addressed before 3.5 release, which is scheduled for next week.
If consumers of packages encounter issues with 3.5 that require them to change their own code, then that's a breaking change and cannot be permitted.
@calebmshafer

**Comment by pmconne** [comment]
@ColinKerr you said "there may be some problems". Please elaborate.
In particular, is there any scenario in which an application's attempt to open an iModel in **read-only** mode will be impacted?

**Comment by ColinKerr** [comment]
> @ColinKerr you said "there may be some problems". Please elaborate.
> In particular, is there any scenario in which an application's attempt to open an iModel in **read-only** mode will be impacted?

The profile upgrade issue cannot happen in readonly mode because it requires a schema import which requires a read/write connection.  

The only issue that can cause readonly mode failures is #2.  We should have a PR for this on the C++ side before the end of the week, @naveedkhan8067 has zeroed in on the problem.

Issues #1 and #3 are limited to schema import scenarios and both could have been avoided if the calling code was written 'correctly'.  Nicks PR should make it easier to write correct code for #3.  We have to investigate ways to make #1 easier but we are not currently planning to have any changes in place for 3.5.

**Comment by pmconne** [comment]
> The only issue that can cause readonly mode failures is https://github.com/iTwin/itwinjs-core/issues/2. We should have a PR for this on the C++ side before the end of the week,

Please provide a link to the issue being addressed.

**Comment by pmconne** [comment]
> We are not tracking this issue in the public backlog. 

#4677

**Comment by calebmshafer** [comment]
@ColinKerr is it feasible to revert the change until we can get the fix in place? If not, we may want to hold the release until that fix is in place.

**Comment by calebmshafer** [comment]
sounds good @ColinKerr.

@nick4598 we can probably abandon this one and either grab the fix from native with a new addon or the rollback of the schema updates. Either way we'll need a new version.

**Pull request closed without merging**